### PR TITLE
Update stacked inline template to support Django's dynamic inline JS

### PR DESCRIPTION
**Pull request by AlexHill**
See stephenmcd/mezzanine#1575

### CODE REVIEW

**Comment by jerivas** [comment]
Looks like I don't have access to this repo, hehe. Here are some CSS fixes for `forms.css` to correctly display the "X" icon on the newly created inlines:

```css
/* Style overrides for the Delete link on dynamic stacked inlines */
.inline-stacked span:last-child {
    background: url('../img/icons/icon-inline_item_tools-deletelink.png') 50% 50% no-repeat;
    float: right;
    height: 15px;
    position: relative;
    text-indent: -9000px;
    width: 20px;
    z-index: 1;
}
.inline-stacked span:last-child:hover {
    background-image: url('../img/icons/icon-inline_item_tools-deletelink-hover.png');
}
.inline-stacked span > a.inline-deletelink {
    display: block;
    height: 100%;
    width: 100%;
}
```

**Comment by AlexHill** [comment]
Thanks Ed! I have updated the PR but can't test right now.

**Comment by stephenmcd** [comment]
> Looks like I don't have access to this repo

Just added you now.

**Comment by jerivas** [comment]
Thanks Steve!

**Comment by jerivas** [comment]
Ok, so now that the style issue has been fixed, the only things not working are the dynamically-added instances of TinyMCE. We can use the `added` callback provided by [Django's `inline.js`](https://github.com/django/django/blob/master/django/contrib/admin/static/admin/js/inlines.js#L141) to init the new instances, but how should this behave if the user swaps out TinyMCE?

**Comment by AlexHill** [comment]
Hmm.

Well, TinyMCE is initialised in [tinymce_setup.js](https://github.com/stephenmcd/mezzanine/blob/master/mezzanine/core/static/mezzanine/js/tinymce_setup.js). We'll need to extract that into a function we can pass to Django's formset plugin as you say. People who want to use CKEditor for example could provide a function of the same name that'll be called for each added inline instead of ours.

**Comment by AlexHill** [comment]
The `added` callback is hard to get to since the APIs we use are the `tabularFormset` and `stackedFormset` functions, which in turn call `formset`. The former don't look for a passed `added` callback. However `formset` triggers the 'formset:added' event, so we can hook into that.

**Comment by jerivas** [comment]
Thanks for the guidance, Alex. I added some generic callbacks that TinyMCE will take advantage of on the Mezzanine side (see https://github.com/stephenmcd/mezzanine/pull/1657). Also found a handful of other bugs which are now fixed.

**Comment by jerivas** [comment]
BTW, Github should allow you to do squash+merge automatically to reduce the amount of commits after we review all of this https://github.com/blog/2141-squash-your-commits

**Comment by AlexHill** [comment]
Awesome Ed.

Assigning the TinyMCE handlers to the window object and then using them in event handlers in the inline JS in the inline/stacked templates seems a bit convoluted, and it means we're dependent on grappelli-safe for this to work. I have a long term goal of making Mezzanine work as well with the default admin as it does with Grappelli, so I'd like to come up with an approach that doesn't add anything to those templates. I'll have a crack tomorrow.

**Comment by jerivas** [comment]
I agree. I guess it would be better to bind the event handler directly in `tinymce_setup.js`. If someone drops TinyMCE in favor of something else, they would just need to implement their own version of the handlers (if needed).

**Comment by jerivas** [comment]
Hey @AlexHill, I have a bit of free time now and I could implement the changes once we settle down on a good way to register the callbacks. Any thoughts?

**Comment by AlexHill** [comment]
Hey @jerivas hope I didn't miss the window. I've put my approach at stephenmcd/mezzanine#1674.

Things to note:
* I'm using TinyMCE's jQuery plugin so we don't have to worry about properly initialising and destroying editor objects.
* We have to use `django.jQuery` rather than the default global `jQuery`, or the event handlers won't work properly. I've modified `jquery.tinymce.min.js` to this end as well.

**Comment by jerivas** [comment]
Finally got around to finishing this up. I removed the callback stuff in favor of the event handlers developed by @AlexHill. I also fixed a bug where the `__prefix__` was not being parsed in instances of `FileField`.

IMO this is ready to merge, but let's have another pair of eyes look at it.

**Comment by jerivas** [comment]
I just noticed that this will not work with Django 1.8, because the events we're listening to were introduced on 1.9 https://github.com/django/django/commit/1335aa2fb9f590d786a5baee1b781377dc3a2ec8. For now the easiest thing to do for those staying with the LTS is to grab a copy of `inlines.js` from 1.9 or later and override it at the project level.

Should Mezzanine backport Django's changes to `inlines.js` to support 1.8?

File: grappelli_safe/templates/admin/edit_inline/stacked.html [hunk 1]
```diff
@@ -47,3 +48,34 @@
...
+        // Tweak the Delete link to resemble the ones defined in template
+        var $deleteLink = $row.find(".inline-deletelink");
+        $deleteLink
+            .appendTo($row.find("ul.inline-item-tools"))
+            .wrap("<li/>")
+            .addClass("deletelink")
+            .attr("title", $deleteLink.text())
+            .text("");
+
+        // Make sure the __prefix__ placeholder is parsed in links
```
Comment (ryneeverett) [inline grappelli_safe/templates/admin/edit_inline/stacked.html]: "Make sure" is a bit vague. Is this section only necessary under certain conditions? If so, do we know what those conditions are?

Comment (jerivas) [inline grappelli_safe/templates/admin/edit_inline/stacked.html]: AFAIK it's necessary only if the inlines contain `FileField` instances, as described in #81. Maybe include a link to that issue?

**Comment by stephenmcd** [comment]
I'm not really across this, but if we're happy with it then I'm happy to merge, LMK.

File: grappelli_safe/templates/admin/edit_inline/stacked.html (hunk 1 above)
Comment (jerivas) [inline grappelli_safe/templates/admin/edit_inline/stacked.html]: I added the comment

**Comment by jerivas** [comment]
Just wanted to say this depends on https://github.com/stephenmcd/mezzanine/pull/1674 (or rather they depend on each other). I'd like to hear once more from @AlexHill or @ryneeverett, on worries if we need to wait for a few more days.

**Comment by jerivas** [comment]
Meant to say "no worries".

**Comment by ryneeverett** [comment]
I'm wondering if we could fix #81 in filebrowser-safe simply by changing `FileBrowser.show('{{ final_attrs.id }}',` -> `FileBrowser.show($(this).attr('id'),` in the custom_field.html template. (It may need to be changed to an onclick handler. I don't know what the context is in `href="javascript`.)

**Comment by jerivas** [comment]
Sounds like a good idea, that way we don't have to couple FileBrowser and Grappelli.

I won't be able to look into it for a few days, so feel free to take care of it (and even push to this branch) if anyone has a chance.

**Comment by ryneeverett** [comment]
See stephenmcd/filebrowser-safe#93. I'd say revert the last couple commits and this is good to go.

**Comment by jerivas** [comment]
I just reverted the `__prefix__` handling. There's a small issue with stephenmcd/filebrowser-safe#93 but after fixing it we should be good to go.

**Comment by jerivas** [comment]
Alright, all issues have been ironed out. I'll go ahead and merge the following:

- This PR where most changes to the inline templates are happening
- stephenmcd/filebrowser-safe#93 to fix instances of `FileField` in inlines
- stephenmcd/mezzanine#1674 to fix instances of `RichTextField` in inlines

**Comment by stephenmcd** [comment]
Thanks Ed!

**Pull request merged**